File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.util.Scanner;

public class Main {

    private static final int[] dx = {1, -1, 0, 0};

    private static final int[] dy = {0, 0, 1, -1};

    private static int[][] graph;

    private static boolean[][] visited;

    private static int maxWater = 0;

    private static int resultCnt = 0;

    public static void main(String[] args) {
        try (Scanner scanner = new Scanner(System.in)) {
            int n = scanner.nextInt();

            graph = new int[n][n];

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    int water = scanner.nextInt();
                    graph[i][j] = water;
                    maxWater = Math.max(maxWater, water);
                }
            }

            for (int i = 0; i <= maxWater; i++) {
                visited = new boolean[n][n];
                int count = 0;
                for (int y = 0; y < n; y++) {
                    for (int x = 0; x < n; x++) {
                        if (!visited[y][x] && graph[y][x] > i) {
                            dfs(x, y, i);
                            count++;
                        }
                    }
                }
                resultCnt = Math.max(resultCnt, count);
            }

            System.out.println(resultCnt);
        }
    }

    private static void dfs(int x, int y, int n) {
        visited[y][x] = true;

        for (int i = 0; i < 4; i++) {
            int cx = x + dx[i];
            int cy = y + dy[i];

            if (cx >= 0 && cy >= 0 && cx < visited.length && cy < visited.length && !visited[cy][cx] && graph[cy][cx] > n) {
                dfs(cx, cy, n);
            }
        }
    }
}

File: 백준/Silver/2468. 안전 영역/안전 영역.kt
import kotlin.math.max

val dx = arrayOf(1, -1, 0, 0)
val dy = arrayOf(0, 0, 1, -1)

fun main() {
    val n = readln().toInt()
    val graph = Array(n) { IntArray(n) }
    lateinit var visited: Array<BooleanArray>

    for (i in 0 until n) {
        val input = readln().split(" ").map { it.toInt() }
        input.forEachIndexed { index, value ->
            graph[i][index] = value
        }
    }

    val max = graph.maxOf { it.max() }
    var maxCount = 1

    for (k in 1..max) {
        var count = 0
        // 물에 잠기는 지역 체크
        val newGraph = graph.copyOf()
        for (i in graph.indices) {
            for (j in graph.indices) {
                if (newGraph[i][j] <= k) {
                    newGraph[i][j] = 0
                }
            }
        }

        visited = Array(n) { BooleanArray(n) }
        for (i in graph.indices) {
            for (j in graph.indices) {
                if (newGraph[i][j] != 0 && !visited[i][j]) {
                    dfs(i, j, newGraph, visited)
                    count++
                }
            }
        }

        maxCount = max(count, maxCount)
    }

    println(maxCount)
}

private fun dfs(
    x: Int,
    y: Int,
    graph: Array<IntArray>,
    visited: Array<BooleanArray>
) {
    visited[x][y] = true

    for (i in 0..3) {
        val dxx = x + dx[i]
        val dyy = y + dy[i]

        if (dxx in graph.indices && dyy in graph.indices && graph[dxx][dyy] != 0 && !visited[dxx][dyy]) {
            dfs(dxx, dyy, graph, visited)
        }
    }
}